File: solutions/python/run-length-encoding/1/run_length_encoding.py
from itertools import groupby

def decode(encoded):
    count = ''
    s = ''
    for k, g in groupby(encoded):
        if k.isnumeric():
            count += str(k)
        elif not count:
            s += str(k)
        else:
            s += str(int(count) * k)
            count = ''
    return s

def encode(input):
    encoded = ''
    for k, g in groupby(input):
        section = sum(1 for _ in g)
        if section > 1:
            encoded += str(section) + k
        else:
             encoded += k
    return encoded

File: solutions/ruby/run-length-encoding/1/run_length_encoding.rb
module RunLengthEncoding
    def self.encode(input)
        input.chars.chunk_while { |prev, curr| prev == curr }
        .map { |chunk| "#{chunk.size if chunk.size > 1}#{chunk.first}" }
        .join
    end

    def self.decode(input)
        input.scan(/(\d*)(\D)/).map do |count, char| 
            char * (count.empty? ? 1 : count.to_i) 
        end.join
    end
end